klogr/klogr.go

```diff
@@ -86,18 +91,31 @@ func copySlice(in []interface{}) []interface{} {
 // Magic string for intermediate frames that we should ignore.
 const autogeneratedFrameName = "<autogenerated>"
 
+// Cached depth of this interface's log functions.
+var framesAtomic int32 // atomic
+
 // Discover how many frames we need to climb to find the caller. This approach
 // was suggested by Ian Lance Taylor of the Go team, so it *should* be safe
 // enough (famous last words).
 //
 // It is needed because binding the specific klogger functions to the
 // logr interface creates one additional call frame that neither we nor
 // our caller know about.
+//
+// This assumes that all logging paths are the same depth from the caller,
+// which should be a reasonable assumption since they are part of the same
+// interface.
 func framesToCaller() int {
+	// Figuring out the current depth is somewhat expensive. Saving the value
+	// amortizes most of that runtime cost.
+	if atomic.LoadInt32(&framesAtomic) != 0 {
+		return int(framesAtomic)
+	}
```

Comment:
This enhancement is unrelated to the logr update, right? A separate PR for it might be better.

Regarding this function: I've not seen any case where it actually encountered an `autogeneratedFrameName`, i.e. the function always returned 1. I wonder whether it is still needed at all.

Reply:
This is a good point.  

Revisiting the original suggstion for this (https://groups.google.com/g/golang-nuts/c/AAcyf5muVPw) we no longer trigger the embedded-struct case in logr, so we probably CAN remove it!

Reply:
ok removing the changes in this section